File: 103-binary-tree-zigzag-level-order-traversal/binary-tree-zigzag-level-order-traversal.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    vector<vector<int>> zigzagLevelOrder(TreeNode* root) {
        vector<vector<int>> ans;
        queue<TreeNode*> q;
        if(root==nullptr) return ans;
        q.push(root);
        int flag = 0;
        while(!q.empty()){
            int s = q.size();
            vector<int> temp(s, 0);
            for(int i=0;i<s;i++){
                TreeNode* node = q.front();
                q.pop();
                if(node->left!=nullptr) q.push(node->left);
                if(node->right!=nullptr) q.push(node->right);
                if(!flag) temp[i] = node->val;
                else temp[s-i-1] = node->val;
            }
            flag = !flag;
            ans.push_back(temp);
        }
        return ans;
    }
};

File: 103-binary-tree-zigzag-level-order-traversal/binary-tree-zigzag-level-order-traversal.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public List<List<Integer>> zigzagLevelOrder(TreeNode root) {
        List<List<Integer>> ans = new ArrayList<>();
        if (root == null)
            return ans;
        Queue<TreeNode> queue = new ArrayDeque<>();
        queue.add(root);
        int cnt = 0;
        while (!queue.isEmpty()) {
            int sz = queue.size();
            List<Integer> sub = new ArrayList<>();
            cnt++;
            for (int i = 0; i < sz; i++) {
                TreeNode node = queue.peek();
                sub.add(node.val);
                if (node.left != null) {
                    queue.add(node.left);
                }
                if (node.right != null) {
                    queue.add(node.right);
                }
                queue.poll();
            }
            if(cnt % 2 == 0) {
                Collections.reverse(sub);
            }
            ans.add(sub);
        }
        return ans;
    }
}